src/Plutarch/Extra/Map.hs

```diff
@@ -1,185 +1,266 @@
+
+ @since 3.2.0
+-}
+pkvPairValue ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData v) =>
+    Term s (PBuiltinPair (PAsData k) (PAsData v) :--> v)
+pkvPairValue = phoistAcyclic $ plam $ \kv -> pfromData (psndBuiltin # kv)
 
--- | @since 1.0.0
-plookup' ::
+{- | Compare two key-value pairs by their keys. Gives 'PTrue' if the key of the
+ first argument pair is less than the key of the second argument pair.
+
+ @since 3.2.0
+-}
+pkvPairLt ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData k, PPartialOrd k) =>
+    Term
+        s
+        ( PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBool
+        )
+pkvPairLt = phoistAcyclic $
+    plam $ \kv kv' ->
+        (pkvPairKey # kv) #< (pkvPairKey # kv')
+
+{- | As 'plookup', but errors when the key is missing.
+
+ @since 3.2.0
+-}
+plookupPartial ::
     forall (k :: S -> Type) (v :: S -> Type) (keys :: KeyGuarantees) (s :: S).
-    (PIsData v, PIsData k, PEq k) =>
+    (PIsData k, PIsData v) =>
     Term s (k :--> PMap keys k v :--> v)
-plookup' = phoistAcyclic $
-    plam $ \x m -> unTermCont $ do
-        res <- pmatchC (plookup # x # m)
-        pure $ case res of
-            PNothing -> ptraceError "plookup': Could not find key."
+plookupPartial = phoistAcyclic $
+    plam $ \k kvs ->
+        pmatch (plookup # k # kvs) $ \case
+            PNothing -> ptraceError "plookupPartial: No value found for key."
             PJust v -> v
 
--- | @since 1.0.0
-pmapFromList ::
-    forall (k :: S -> Type) (v :: S -> Type) (keys :: KeyGuarantees) (s :: S).
-    (PIsData k, PIsData v) =>
-    [(Term s k, Term s v)] ->
-    Term s (PMap keys k v)
-pmapFromList = pcon . PMap . foldl' go (pcon PNil)
+{- | Given a 'Foldable' of key-value pairs, construct an unsorted 'PMap'.
+ Performs linearly with respect to its argument.
+
+ = Note
+
+ If there are duplicate keys in the input, the /last/ key will \'win\' in a
+ lookup.
+
+ @since 3.2.0
+-}
+pmapFromFoldableUnsorted ::
+    forall (k :: S -> Type) (v :: S -> Type) (f :: Type -> Type) (s :: S).
+    (Foldable f, PIsData k, PIsData v) =>
+    f (Term s k, Term s v) ->
+    Term s (PMap 'Unsorted k v)
+pmapFromFoldableUnsorted = pcon . PMap . foldl' go (pcon PNil)
   where
-    go acc (k, v) = unTermCont $ do
-        k' <- pletC (pdata k)
-        v' <- pletC (pdata v)
-        p <- pletC (ppairDataBuiltin # k' # v')
-        pure . pcon . PCons p $ acc
+    go ::
+        forall (s' :: S).
+        Term s' (PBuiltinList (PBuiltinPair (PAsData k) (PAsData v))) ->
+        (Term s' k, Term s' v) ->
+        Term s' (PBuiltinList (PBuiltinPair (PAsData k) (PAsData v)))
+    go acc (key, val) =
+        pcon . PCons (ppairDataBuiltin # pdata key # pdata val) $ acc
 
-{- | Get the keys of a given map, the order of the keys is preserved.
+{- | Given a 'Foldable' of (not necessarily sorted) key-value pairs, construct a
```

Comment:
It'd be nice to mention here a duplicate entry will get disposed. I think `pinsert` replace the old one with the new one, so it has quite a different behavior as compared to its unsafe counterpart.